CARDS-1902: Add support for using Cloud-IAM.com as a SAML IdP for testing purposes

IntegralProgrammer opened the pull request:
This PR adds, to the `start_cards.sh` script and the `generate_compose_yaml.py` script, support for using a pre-configured free instance of _Keycloak_ on _Cloud-IAM.com_ for testing CARDS authentication via SAML.

### Testing

- Build this `CARDS-1902` branch, and Docker image, with `mvn clean install -Pdocker`.

#### Testing `start_cards.sh`

- Download the `samlKeystore.p12` file pinned to the **#cards-dev** Slack channel and save it in the root `cards` directory.
- `./start_cards.sh --saml --cloud-iam_demo --dev -p 8181` (you don't necessarily need to use port `8181`, just something other than `8080`) 
- Visit http://localhost:8080 and you should be able to login either as `alice@cloud-iam.demo` or `bob@cloud-iam.demo` with the password provided in the **#cards-dev** Slack channel. By default, `alice@cloud-iam.demo` and `bob@cloud-iam.demo` do not belong to the `TrustedUsers` group and therefore they will need to be added to that group by `admin` in order to see any data stored in CARDS.

#### Testing `generate_compose_yaml.py`

- Download the `samlKeystore.p12` file pinned to the **#cards-dev** Slack channel and save it in the `compose-cluster` directory.
- Enter the `compose-cluster` directory, clean up any previous configurations with `./cleanup.sh`, and run `python3 generate_compose_yaml.py --dev_docker_image --composum --oak_filesystem --saml --saml_cloud_iam_demo --server_address localhost:8080`.
- `docker-compose build`
- `docker-compose up -d`
- Visit http://localhost:8080 and you should be able to login either as `alice@cloud-iam.demo` or `bob@cloud-iam.demo` with the password provided in the **#cards-dev** Slack channel. Just as is the case with `start_cards.sh`, by default, `alice@cloud-iam.demo` and `bob@cloud-iam.demo` do not belong to the `TrustedUsers` group and therefore they will need to be added to that group by `admin` in order to see any data stored in CARDS.

IntegralProgrammer committed:
CARDS-1902: Add support for using Cloud-IAM.com as a SAML IdP for testing purposes

Add the SAML domain cloud-iam.demo and make it available through the --cloud-iam_demo start_cards.sh flag

IntegralProgrammer committed:
CARDS-1902: Add support for using Cloud-IAM.com as a SAML IdP for testing purposes

Add the setup_saml_cloud-iam_demo.sh script (and associated file changes) to enable SAML login via Cloud-IAM.com for a running CARDS instance

IntegralProgrammer committed:
CARDS-1902: Add support for using Cloud-IAM.com as a SAML IdP for testing purposes

If start_cards.sh is called with the --cloud-iam_demo flag, call the setup_saml_cloud-iam_demo.sh script during CARDS startup.

IntegralProgrammer committed:
CARDS-1902: Add support for using Cloud-IAM.com as a SAML IdP for testing purposes

Add support for the --keycloak-endpoint flag to keycloak_headermod_http_proxy.js so that an arbitrary Keycloak endpoint can be used while still respecting the default value of http://localhost:8484/auth/realms/myrealm/protocol/saml

IntegralProgrammer committed:
CARDS-1902: Add support for using Cloud-IAM.com as a SAML IdP for testing purposes

If start_cards.sh is called with the --saml flag, start the keycloak_headermod_http_proxy.js proxy

IntegralProgrammer committed:
CARDS-1902: Add support for using Cloud-IAM.com as a SAML IdP for testing purposes

If start_cards.sh is called with the --saml flag and the --cloud-iam_demo flag, start the keycloak_headermod_http_proxy.js proxy with the --keycloak-endpoint=https://lemur-15.cloud-iam.com/auth/realms/cards-saml-test/protocol/saml argument

IntegralProgrammer committed:
CARDS-1902: Add support for using Cloud-IAM.com as a SAML IdP for testing purposes

Change the message_started_cards message in start_cards.sh to reflect presence of the keycloak_headermod_http_proxy.js proxy

IntegralProgrammer committed:
CARDS-1902: Add support for using Cloud-IAM.com as a SAML IdP for testing purposes

Add support for the --cards-port flag to keycloak_headermod_http_proxy.js to specify the port which CARDS is listening on (which is not always 8080)

IntegralProgrammer committed:
CARDS-1902: Add support for using Cloud-IAM.com as a SAML IdP for testing purposes

Add support for the --listen-port flag for the keycloak_headermod_http_proxy.js so that it can listen on an arbitrary TCP port - not just 9090

IntegralProgrammer committed:
CARDS-1902: Add support for using Cloud-IAM.com as a SAML IdP for testing purposes

When start_cards.sh is being used with the --cloud-iam_demo flag, keycloak_headermod_http_proxy.js must listen on port 8080 in order to match what is configured on the Cloud-IAM.com service

IntegralProgrammer committed:
CARDS-1902: Add support for using Cloud-IAM.com as a SAML IdP for testing purposes

The setup_saml_cloud-iam_demo.sh script is no longer need, instead we specify the Cloud-IAM.com SAML configuration in the cloud-iam-demo-saml feature.json file

IntegralProgrammer committed:
CARDS-1902: Add support for using Cloud-IAM.com as a SAML IdP for testing purposes

Add Docker support for the Cloud-IAM.com SAML authentication demo via the --saml_cloud_iam_demo generate_compose_yaml.py flag

IntegralProgrammer committed:
CARDS-1902: Add support for using Cloud-IAM.com as a SAML IdP for testing purposes

When generate_compose_yaml.py is used with the --saml_cloud_iam_demo flag, verify the SHA256 hash of samlKeystore.p12 and display a warning if it does not match its expected value

sdumitriu commented on modules/cloud-iam-demo-saml/pom.xml:
```diff
@@ -0,0 +1,51 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I think we should move all the saml-related modules under `/modules/saml`, there are too many already.
- `cloud-iam-demo-saml`
- `fetch-requires-saml-login`
- `keycloakdemo-saml`
- `saml-support`
- `uhn-saml`

IntegralProgrammer replied:
I've partially addressed this in commit https://github.com/data-team-uhn/cards/pull/1156/commits/0e9ff6a409016fa38e39b519d453339ce0cd0daa and commit https://github.com/data-team-uhn/cards/pull/1156/commits/e26eed9b58ec9f09039133c0da38ee9c3d2c68c8 by moving the `cloud-iam-demo-saml` module into a new `modules/saml` directory.

I agree that `fetch-requires-saml-login`, `keycloakdemo-saml`, `saml-support`, and `uhn-saml` should also be moved under the `modules/saml` directory but let's do that in a different PR since it's not within the scope of CARDS-1902.

IntegralProgrammer committed:
CARDS-1902: Add support for using Cloud-IAM.com as a SAML IdP for testing purposes

Move the cloud-iam-demo-saml module from under modules/ to under modules/saml

In the future, other SAML-related modules (eg. keycloakdemo-saml, uhn-saml) will be moved into this directory

IntegralProgrammer committed:
CARDS-1902: Add support for using Cloud-IAM.com as a SAML IdP for testing purposes

Move the cloud-iam-demo-saml module from under modules/ to under modules/saml

In the future, other SAML-related modules (eg. keycloakdemo-saml, uhn-saml) will be moved into this directory

Change parent in modules/saml/pom.xml from cards-parent to cards-modules

sdumitriu commented on modules/saml/cloud-iam-demo-saml/pom.xml:
```diff
@@ -0,0 +1,51 @@
+
+  <parent>
+    <groupId>io.uhndata.cards</groupId>
+    <artifactId>cards-modules</artifactId>
```
```suggestion
    <artifactId>cards-saml-modules</artifactId>
```

sdumitriu committed:
[misc] Update submodule's parent
